File: src/shared/token.ts
import { Location } from "./location";

export const WORD_CHARS =
    "QWERTYUIOPASDFGHJKLZXCVBNM"  // Uppercase letters
  + "qwertyuiopasdfghjklzxcvbnm"  // Lowercase letters
  + "~!@#$%^&*()_-+=|;:,./<>?"    // Special characters
  + "0123456789";                 // Numbers

export const KEYWORDS = new Set([
  "proc", "macro", "const", "include", "end", "if", "if*", "else", "while", "do"
]);

export enum Tokens {
  Int = "<int>",
  Str = "<str>",
  Char = "<char>",
  Boolean = "<boolean>",
  Word = "<word>",

  Proc = "proc",
  Macro = "macro",
  Const = "const",
  Include = "include",
  End = "end",
  If = "if",
  ChainedIf = "if*",
  Else = "else",
  While = "while",
  Do = "do",

  // Special token appended at the end of the file
  EOF = "EOF"
}

export interface Token {
  kind: Tokens;
  loc: Location;
  value?: any;
}

File: src/parser/token.rs
use crate::shared::location::Span;

#[derive(Clone, PartialEq, Eq, Debug)]
pub enum Tokens {
    Str(String),
    Int(i64),
    Bool(bool),
    Word(String),

    Macro,
    Include,

    Proc,
    Do,
    End,
}

#[derive(Clone, PartialEq, Eq, Debug)]
pub enum TokenKind {
    Str,
    Int,
    Bool,
    Word,
    Macro,
    Include,
    Proc,
    Do,
    End
}

#[derive(Clone, PartialEq, Eq, Debug)]
pub struct Token {
    pub kind: Tokens,
    pub span: Span,
}

impl Token {
    pub fn new(kind: Tokens, span: Span) -> Self {
        Self { kind, span }
    }
}

impl Tokens {
    pub fn to_kind(&self) -> TokenKind {
        match self {
            Tokens::Str(_) => TokenKind::Str,
            Tokens::Int(_) => TokenKind::Int,
            Tokens::Bool(_) => TokenKind::Bool,
            Tokens::Word(_) => TokenKind::Word,
            Tokens::Macro => TokenKind::Macro,
            Tokens::Include => TokenKind::Include,
            Tokens::Proc => TokenKind::Proc,
            Tokens::Do => TokenKind::Do,
            Tokens::End => TokenKind::End,
        }
    }

    pub fn to_str(&self) -> &'static str {
        self.to_kind().to_str()
    }
}

impl TokenKind {
    pub fn to_str(&self) -> &'static str {
        match self {
            TokenKind::Str => "<str>",
            TokenKind::Int => "<int>",
            TokenKind::Bool => "<bool>",
            TokenKind::Word => "<word>",
            TokenKind::Macro => "macro",
            TokenKind::Include => "include",
            TokenKind::Proc => "proc",
            TokenKind::Do => "do",
            TokenKind::End => "end"
        }
    }
}